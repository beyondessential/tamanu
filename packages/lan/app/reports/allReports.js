export * from './returnVivax';
export * from './diagnosesByVillage';
export * from './anemiaVivaxCodiagnoses';

export * from './visits';
