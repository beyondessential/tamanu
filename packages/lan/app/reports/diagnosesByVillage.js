export const diagnosesByVillageReport = async (db, { startDate, endDate, village }) => {
  const baseDiagnoses = db
    .objects('patientDiagnosis')
    .filtered(
      'date >= $0 AND date <= $1 AND visit.patient.village.name = $2',
      startDate,
      endDate,
      village,
    );

  const counts = {};

  baseDiagnoses.forEach(d => {
    if (!counts[d.diagnosis._id]) {
      counts[d.diagnosis._id] = {
        count: 0,
        diagnosis: d.diagnosis.name,
      };
    }
    counts[d.diagnosis._id].count += 1;
  });

  return {
    headers: ['diagnosis', 'count'],
    rowData: Object.values(counts),
  };
};
