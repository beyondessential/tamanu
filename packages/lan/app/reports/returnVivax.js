import { formatPatientInfo } from './utils';

export const returnVivaxReport = {
  title: 'Returning cases of malaria (vivax)',
  parameters: {
    startDate: 'date',
    endDate: 'date',
  },
  run: async (db, { startDate, endDate }) => {
    const search = 'vivax';
    const baseDiagnoses = db
      .objects('patientDiagnosis')
      .filtered('diagnosis.name CONTAINS[c] $0', search)
      .filtered('date >= $0 AND date <= $1', startDate, endDate)
      .sorted('date');

    const counts = {};
    const repeats = {};
    baseDiagnoses.forEach(x => {
      const patientId = x.visit[0].patient[0]._id;
      counts[patientId] = (counts[patientId] || 0) + 1;
      if (counts[patientId] > 1) {
        repeats[patientId] = x;
      }
    });

    const rowData = Object.values(repeats).map(diagnosis => ({
      ...formatPatientInfo(diagnosis.visit[0].patient[0]),
      diagnosis: diagnosis.diagnosis.name,
      date: diagnosis.date,
    }));

    return {
      headers: ['id', 'name', 'village', 'diagnosis', 'lastDiagnosed'],
      rowData,
    };
  },
};
