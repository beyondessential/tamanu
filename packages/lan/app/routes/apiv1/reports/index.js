import express from 'express';
import { createAdmissionsReport } from './admissions';

export const reports = express.Router();

reports.post('/admissions', createAdmissionsReport);
