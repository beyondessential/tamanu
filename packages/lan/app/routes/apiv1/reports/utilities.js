export function generateReportFromQueryData(queryData, columnTemplate) {
  return [columnTemplate.map(c => c.title), ...queryData.map(r => columns.map(c => c.accessor(r)))];
}
