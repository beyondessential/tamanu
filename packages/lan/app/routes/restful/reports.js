import express from 'express';
import moment from 'moment';

import { objectToJSON } from '../../utils';
import { fillDateRange } from 'Shared/utils/fillDateRange';

export const reportsRoutes = express.Router();

reportsRoutes.get('/diagnosesByWeek', (req, res) => {
  const { db, query } = req;
  const { 
    startDate = '', 
    endDate = '',
    diagnoses = '',
  } = query;

  const ranges = fillDateRange(startDate, endDate, 'week');
  const columns = ranges.map(x => x.format('DD/MM/YYYY'));

  const results = diagnoses.split(',').map(_id => {
    const diagnosis = db.objectForPrimaryKey('diagnosis', _id);
    const diagnoses = db.objects('patientDiagnosis')
      .filtered('diagnosis._id = $0', _id);

    const countForDates = (start, end) => diagnoses
      .map(diagnosis => moment(diagnosis.date))
      .filter(date => date >= start && date <= end)
      .length;

    return {
      key: _id,
      formatted: diagnosis.name,
      values: ranges.map(r => countForDates(r, moment(r).add(1, 'week'))),
    }
  });

  const meta = {
    title: "Diagnosis",
    columns,
  };

  res.send({ results, meta });
});
