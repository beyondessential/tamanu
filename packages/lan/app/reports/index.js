import XLSX from 'xlsx';
import moment from 'moment';

import * as allReports from './allReports';

const tabulate = ({ headers, rowData }) => [
  headers,
  ...rowData.map(row => headers.map(h => row[h])),
];

const writeToExcel = async (path, data) => {
  const sheet = XLSX.utils.aoa_to_sheet(tabulate(data));

  const book = XLSX.utils.book_new();
  XLSX.utils.book_append_sheet(book, sheet, 'values');

  XLSX.writeFile(book, path);
};

export const getAllReports = async () => {
  return Object.entries(allReports)
    .map(([id, { title, parameters }]) => ({ id, title, parameters }));
};

export const generateReport = async (db, reportName, userParams) => {
  const report = allReports[reportName];
  if (!report) {
    throw new Error('No such report');
  }

  const params = {
    startDate: moment(userParams.endDate)
      .subtract(1, 'month')
      .toDate(),
    endDate: moment().toDate(),
    ...userParams,
  };

  const data = await report.run(db, params);

  // const date = moment().format('YYYY-MM-DD');
  // const filename = `${date}_${reportName}.xlsx`;

  // await writeToExcel(filename, data);

  return {
    metadata: {
      report: report.title,
      generated: moment().toDate(),
      ...userParams,
    },
    data,
  };
};
