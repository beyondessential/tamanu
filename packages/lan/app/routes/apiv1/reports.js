import express from 'express';
import asyncHandler from 'express-async-handler';
import * as reportUtils from 'shared/reports';
import { REPORT_STATUSES, REPORT_DATE_RANGE_LABELS } from 'shared/constants';
import { createNamedLogger } from 'shared/services/logging/createNamedLogger';
import { assertReportEnabled } from '../../utils/assertReportEnabled';

const FACILITY_REPORT_LOG_NAME = 'FacilityReport';

export const reports = express.Router();

reports.get(
  '/$',
  asyncHandler(async (req, res) => {
    req.flagPermissionChecked();
    const { models, user, ability } = req;
    const { UserLocalisationCache, ReportDefinition } = models;

    const localisation = await UserLocalisationCache.getLocalisation({
      where: { userId: user.id },
      order: [['createdAt', 'DESC']],
    });

    const disabledReports = localisation?.disabledReports || [];
    const availableReports = reportUtils.REPORT_DEFINITIONS.filter(
      ({ id }) =>
        !disabledReports.includes(id) && ability.can('run', reportUtils.REPORT_OBJECTS[id]),
    ).map(report => ({ ...report, legacyReport: true }));

    const reportsDefinitions = await ReportDefinition.findAll({
      include: [
        {
          model: models.ReportDefinitionVersion,
          as: 'versions',
          where: { status: REPORT_STATUSES.PUBLISHED },
        },
      ],
      order: [['versions', 'version_number', 'DESC']],
    });

    const dbReports = reportsDefinitions.map(r => {
      // Get the latest report definition version by getting the first record from the ordered list
      const version = r.versions[0];

      return {
        id: version.id,
        name: r.name,
        dataSourceOptions: version.queryOptions.dataSources,
        filterDateRangeAsStrings: true,
        dateRangeLabel:
          version.queryOptions.dateRangeLabel ||
          REPORT_DATE_RANGE_LABELS[version.queryOptions.defaultDateRange],
        parameters: version.getParameters(),
        version: version.versionNumber,
      };
    });

    res.send([...availableReports, ...dbReports]);
  }),
);

reports.post(
  '/:reportId',
  asyncHandler(async (req, res) => {
    const {
      db,
      models,
      body: { parameters },
      user,
      params,
      getLocalisation,
    } = req;
    // Permissions are checked per report module
    req.flagPermissionChecked();
    const { reportId } = params;
    const facilityReportLog = createNamedLogger(FACILITY_REPORT_LOG_NAME, {
      userId: user.id,
      reportId,
    });
    const localisation = await getLocalisation();
    assertReportEnabled(localisation, reportId);

    const reportModule = await reportUtils.getReportModule(reportId, models);

    if (!reportModule) {
      const message = 'Report module not found';
      facilityReportLog.error(message);
      res.status(400).send({ error: { message } });
      return;
    }

    req.checkPermission('read', reportModule.permission);

    try {
      facilityReportLog.info('Running report', { parameters });
      const excelData = await reportModule.dataGenerator({ sequelize: db, models }, parameters);
      facilityReportLog.info('Report run successfully');
      res.send(excelData);
    } catch (e) {
      facilityReportLog.error('Report module failed to generate data', {
        stack: e.stack,
      });
      res.status(400).send({
        error: {
          message: e.message,
        },
      });
    }
  }),
);
