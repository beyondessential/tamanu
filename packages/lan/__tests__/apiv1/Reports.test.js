import * as reportsUtils from 'shared/reports';
import { fake } from 'shared/test-helpers';
import { createTestContext } from '../utilities';

describe('Reports', () => {
  let baseApp = null;
  let ctx;

  beforeAll(async () => {
    ctx = await createTestContext();
    baseApp = ctx.baseApp;
  });
  afterAll(() => ctx.close());

  describe('database defined reports', () => {
    let adminApp = null;
    let reportDefinition = null;
    let user = null;
    beforeAll(async () => {
      adminApp = await baseApp.asRole('admin');
      const { models } = ctx;
      user = await models.User.create({
        ...fake(models.User),
        email: 'test@tamanu.io',
      });
      await models.ReportDefinition.create({
        ...fake(models.ReportDefinition),
        name: 'test-report',
      });
      reportDefinition = await models.ReportDefinitionVersion.create({
        versionNumber: 1,
        status: 'published',
        userId: user.id,
        queryOptions: { parameters: [{ parameterField: 'EmailField', name: 'email' }] },
        query: 'SELECT * from users WHERE email LIKE :email;',
      });
    });
    it('should run a simple database defined report', async () => {
      const response = await adminApp.post(`/v1/reports/${reportDefinition.id}`);
      expect(response).toHaveSucceeded();
      // There will be more than one user because of the app context
      expect(response.body.length).toBeGreaterThan(1);
    });
    it('should apply filters on a database defined report', async () => {
      const response = await adminApp.post(`/v1/reports/${reportDefinition.id}`).send({
        parameters: {
          email: user.email,
        },
      });
      expect(response).toHaveSucceeded();
      expect(response.body.length).toEqual(2);
      const headerRow = response.body[0];
      const firstRow = response.body[1];
      expect(headerRow[0]).toEqual('id');
      expect(headerRow[1]).toEqual('email');
      expect(firstRow[1]).toEqual(user.email);
    });
  });

  describe('post', () => {
    let app = null;
    beforeAll(async () => {
      app = await baseApp.asRole('practitioner');
    });

    it('should reject reading a report with insufficient permissions', async () => {
      const noPermsApp = await baseApp.asRole('base');
      const result = await noPermsApp.post('/v1/reports/incomplete-referrals', {});
      expect(result).toBeForbidden();
    });
    it('should fail with 400 and message if report module is not found', async () => {
      jest.spyOn(reportsUtils, 'getReportModule').mockResolvedValue(null);
      const res = await app.post('/v1/reports/invalid-report', {});
      expect(res).toHaveStatus(400);
      expect(res.body).toEqual({ error: { message: 'Report module not found' } });
    });
    it('should fail with 400 and error message if dataGenerator encounters error', async () => {
      jest.spyOn(reportsUtils, 'getReportModule').mockResolvedValue({
        dataGenerator: jest.fn().mockRejectedValue(new Error('test-error-message')),
        permission: 'Patient', // just need any valid permission
      });
      const res = await app.post('/v1/reports/incomplete-referrals', {});
      expect(res).toHaveStatus(400);
      expect(res.body).toEqual({ error: { message: 'test-error-message' } });
    });
  });
});
