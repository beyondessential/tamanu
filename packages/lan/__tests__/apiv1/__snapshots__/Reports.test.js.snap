// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reports post permissions should be able to run permitted db reports 1`] = `
[
  [
    "test_column",
  ],
  [
    2,
  ],
]
`;
